features/step_definitions/mention_steps.js

```diff
@@ -0,0 +1,125 @@
+        },
+    ];
+
+    // Create a public Note with the mention (uses defaults: to='as:Public')
+    const object = await createObject('Note', actor, `Hello ${mention}`, tags);
+    const activity = await createActivity('Create', object, actor);
+
+    await fetchActivityPub('https://self.test/.ghost/activitypub/inbox/index', {
+        method: 'POST',
+        headers: {
+            'Content-Type': 'application/ld+json',
+        },
+        body: JSON.stringify(activity),
+    });
+
+    this.mentionId = object.id;
+    this.activities['Create(Note)'] = activity;
+    this.objects.Note = object;
+});
+
+When('{string} sends us a private mention', async function (actorName) {
+    const actor = this.actors[actorName];
+    if (!actor) {
+        throw new Error(
+            `Actor ${actorName} not found - did you forget a step?`,
+        );
+    }
+
+    // Use the existing local actor (Us) from the test context
+    const localActor = this.actors.Us;
+    if (!localActor) {
+        throw new Error('Local actor (Us) not found in test context');
+    }
+
+    const mention = `@${localActor.preferredUsername}@self.test`;
+    const tags = [
+        {
+            type: 'Mention',
+            name: mention,
+            href: localActor.id,
+        },
+    ];
+
+    // Create a Note with the mention
+    const object = await createObject('Note', actor, `Hello ${mention}`, tags);
+
+    // Make it private by addressing it only to the mentioned user
+    object.to = localActor.id;
+    object.cc = [];
+
+    const activity = await createActivity('Create', object, actor);
+    activity.to = localActor.id;
+    activity.cc = [];
+
+    await fetchActivityPub('https://self.test/.ghost/activitypub/inbox/index', {
+        method: 'POST',
+        headers: {
+            'Content-Type': 'application/ld+json',
+        },
+        body: JSON.stringify(activity),
+    });
+
+    this.mentionId = object.id;
+    this.activities['Create(Note)'] = activity;
+    this.objects.Note = object;
+});
+
+Then('the mention is in our notifications', async function () {
+    if (!this.mentionId) {
+        throw new Error(
+            'You need to call a step which creates a mention before this',
+        );
+    }
+
+    const found = await waitForItemInNotifications(this.mentionId);
+    assert(found, `Expected mention ${this.mentionId} to be in notifications`);
+});
+
+Then('the mention is not in our notifications', async function () {
+    if (!this.mentionId) {
+        throw new Error(
+            'You need to call a step which creates a mention before this',
+        );
+    }
+
+    try {
+        await waitForItemInNotifications(this.mentionId);
+        assert.fail(
+            `Expected mention ${this.mentionId} to not be in notifications`,
+        );
+    } catch (error) {
+        assert.equal(
+            error.message,
+            `Max retries reached when waiting on item in notifications`,
+        );
+    }
+});
```

Comment:
I'm not sure about this - this step is always going to take ~7.5s to run (waiting for the max timeout)

I get why we need to do this, but it doesn't feel great. Would we be better just keeping the happy path at the the e2e level, and then testing the non-happy paths at the unit (or integration) level?

Also, semantically, not receiving a private mention is not really a feature, more an implementation detail (because we don't support private mention) so we probably don't need to document it via cucumber?

What do you think?

Reply:
Yeah that's fair! I wanted to have a e2e test here to reproduce the actual bug that was reported - a private mention that is treated as public content - but agree the negative e2e test is slow. 

Let me see if I move this test to unit / integratio